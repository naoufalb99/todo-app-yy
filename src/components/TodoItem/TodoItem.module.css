.box {
    border: 2px solid var(--primary-light);
    padding: 1em;
    border-radius: 3px;
    cursor: pointer;
    margin-bottom: 1em;
    display: flex;
    align-items: center;
    gap: 1em;
}

.checkbox {
    width: 1.2em;
    height: 1.2em;
    border: 0.25em solid var(--checkbox-color);
    border-radius: 99rem;
}

.checkbox.completed {
    background-color: var(--checkbox-color);
}

.box.completed {
    text-decoration: line-through;
}
