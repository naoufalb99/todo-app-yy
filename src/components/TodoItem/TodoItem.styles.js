const styles = {
    box: {
        border: '2px solid var(--primary-light)',
        padding: '1em',
        borderRadius: '3px',
        cursor: 'pointer',
        marginBottom: '1em',
        display: 'flex',
        alignItems: 'center',
        gap: '1em',
        textDecoration: ({ completed }) =>
            completed ? 'line-through' : 'none',
    },
    checkbox: {
        width: '1.2em',
        height: '1.2em',
        border: ({ color }) => `0.25em solid ${color}`,
        borderRadius: '99rem',
        backgroundColor: ({ completed, color }) =>
            completed ? color : 'transparent',
    },
};

export default styles;
