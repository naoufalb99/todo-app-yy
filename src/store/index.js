import { createStore, applyMiddleware, compose, combineReducers } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage/session';

import todosReducer from './todos';

const reducers = combineReducers({
    todos: todosReducer,
});

const persistedReducer = persistReducer({ key: 'todo-app', storage }, reducers);

export const configureStore = (initialState = {}) => {
    const middlewares = [];
    const enhancers = [applyMiddleware(...middlewares)];

    const store = createStore(
        persistedReducer,
        initialState,
        compose(...enhancers)
    );
    const persistor = persistStore(store);

    return { store, persistor };
};
