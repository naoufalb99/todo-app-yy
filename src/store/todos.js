const initialState = [
    {
        id: 'id-1',
        text: 'Consider using Redux',
        completed: false,
    },
];

const createActionName = (name) => `todos/${name}`;

const ADD_TODO = createActionName('ADD_TODO');
const TOGGLE_TODO = createActionName('TOGGLE_TODO');

const reducer = (state = initialState, action) => {
    switch (action.type) {
        case ADD_TODO:
            return [
                ...state,
                {
                    id: action.id,
                    text: action.text,
                    completed: action.completed,
                },
            ];
        case TOGGLE_TODO:
            return state.map((todo) =>
                todo.id === action.id
                    ? { ...todo, completed: !todo.completed }
                    : todo
            );
        default:
            return state;
    }
};

// Action creators
const addTodo = (id, text, completed) => ({
    type: ADD_TODO,
    id,
    text,
    completed,
});
const toggleTodo = (id) => ({ type: TOGGLE_TODO, id });

// Selectors
const getTodos = (state) => state.todos;
const getCompletedTodos = (state) =>
    getTodos(state).filter((todo) => todo.completed);
const getUncompletedTodos = (state) =>
    getTodos(state).filter((todo) => !todo.completed);

export {
    reducer as default,
    addTodo,
    toggleTodo,
    getCompletedTodos,
    getUncompletedTodos,
};
