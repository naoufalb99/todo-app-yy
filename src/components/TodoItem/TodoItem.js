import clsx from 'clsx';
import { useDispatch } from 'react-redux';
import { TODO_CHECKBOX_COLORS } from '../../constants';
import { toggleTodo } from '../../store/todos';
import { getRandomElementFromArray } from '../../utils/array';
import styles from './TodoItem.module.css';

function TodoItemCheckbox({ color, completed }) {
    return (
        <div
            className={clsx(styles.checkbox, completed && styles.completed)}
            style={{
                '--checkbox-color': color,
            }}
        />
    );
}

function TodoItem({ completed, id, text }) {
    const disptach = useDispatch();

    const handleToogle = () => {
        disptach(toggleTodo(id));
    };

    return (
        <div
            className={clsx(styles.box, completed && styles.completed)}
            onClick={handleToogle}
        >
            <TodoItemCheckbox
                color={getRandomElementFromArray(TODO_CHECKBOX_COLORS)}
                completed={completed}
            />
            {text}
        </div>
    );
}

export default TodoItem;
