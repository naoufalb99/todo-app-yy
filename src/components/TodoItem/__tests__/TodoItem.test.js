import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import TodoItem from '../TodoItem';
import { toggleTodo } from '../../../store/todos';

jest.mock('../../../utils/array', () => ({
    getRandomElementFromArray(array) {
        return array[0];
    },
}));

const mockStore = configureStore();

const initialState = {
    todos: [
        {
            id: 1,
            text: 'Learn React',
            completed: false,
        },
    ],
};

const todo = initialState.todos[0];

describe('<TodoItem />', () => {
    let mockStore;

    beforeEach(() => {
        mockStore = configureStore();
    });

    it('should match a snapshot', () => {
        const store = mockStore(initialState);

        const wrapper = mount(
            <Provider store={store}>
                <TodoItem
                    id={todo.id}
                    text={todo.text}
                    completed={todo.completed}
                />
            </Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should dispatch TOGGLE_TODO', () => {
        const store = mockStore(initialState);

        const wrapper = mount(
            <Provider store={store}>
                <TodoItem
                    id={todo.id}
                    text={todo.text}
                    completed={todo.completed}
                />
            </Provider>
        );

        wrapper.find('TodoItem > div').simulate('click');

        expect(store.getActions()[0]).toEqual(toggleTodo(todo.id));
    });
});
