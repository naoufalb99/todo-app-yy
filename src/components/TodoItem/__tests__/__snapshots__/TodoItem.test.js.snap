// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem /> should match a snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <TodoItem
    completed={false}
    id={1}
    text="Learn React"
  >
    <div
      className="box-0-2-1 box-d0-0-2-3"
      onClick={[Function]}
    >
      <TodoItemCheckbox
        color="#F0473E"
        completed={false}
      >
        <div
          className="checkbox-0-2-2 checkbox-d3-0-2-6"
        />
      </TodoItemCheckbox>
      Learn React
    </div>
  </TodoItem>
</Provider>
`;
